### ISSUE TITLE
The "Select an input method" button is missing in ibus preferences>input method

### ISSUE THREAD

**Issue by mkniloy** [elementary/switchboard-plug-keyboard#231]
![Screenshot from 2019-08-01 14 48 28](https://user-images.githubusercontent.com/53095171/62279504-0f705b00-b46c-11e9-9b4d-a556ce9d4b33.png)

The "Select an input method" button is missing in ibus preferences>input method in my device

**Comment by ryonakano** [comment]
AFAIK there's no such button in the current version of IBus Preferences, so unexistence of that button is not an issue.

**Comment by mkniloy** [comment]
Then how can I type in Bengali? I tried using the 'ctrl+space' shortcut...but even if i select Bengali input..the language stays English...

**Comment by ryonakano** [comment]
I think you don't have to use IBus, as far as seeing your screenshot you've already added English and Bengali keyboard layout, so simply switching to the Bengali keyboard layout by `Alt + Shift` key should allow you to type in Bengali.

**Comment by mkniloy** [comment]
The default Bengali keyboard layout doesn't match my physical keyboard's layout. I was trying to use the 'Bijoy(Bengali)' layout...because my physical keyboard has 'Bijoy(Bengali)' layout...

**Comment by ryonakano** [comment]
Oh, I understand. I searched for Bijoy keyboard and found [an interesting article](https://www.compromath.com/2017/06/install-bijoy-ekushe-to-linuxubuntu.html) that shows how to use Bijoy keyboard in Ubuntu (Unity). Are you trying to add Bijoy keyboard in that way?

**Comment by mkniloy** [comment]
I've already tried this process shown in this video (https://www.youtube.com/watch?v=ihkNsNwTcOw). As you can see on my photo that I've added 'Bengali - bijoyClassic'(2nd one) inside Input Method Settings. But when I use ctrl+space and select bijoyClassic...the language still remains English. I think it's a bug.

**Comment by ryonakano** [comment]
Thank you for telling a reference, I checked it and enabled bijoyClassic and bijoyUnicode engines in my system. I'm not sure what Bengali characters look like but they seem to work for me.

Could you check if you can type Bengali characters with bijoyUnicode?
(I'm not sure the difference of bijoyClassic and bijoyUnicode, so please let me know if what you need is bijoyClassic, not bijoyUnicode)

**Comment by mkniloy** [comment]
![Screenshot from 2019-08-03 14 53 29](https://user-images.githubusercontent.com/53095171/62410495-572be980-b608-11e9-90cf-506be5cccfb4.png)

No, I can't even use bijoyUnicode...this pop up is telling me to set the Keyboard Input Method to 'ibus'...but the "select an input method" button is missing...

**Comment by ryonakano** [comment]
Ah, that dialog is shown because IBus daemon is not running on your system.

![Screenshot from 2019-08-03 19-20-04](https://user-images.githubusercontent.com/26003928/62410667-ded32180-b623-11e9-931f-5f90505b9c37.png)

Open "Startup" tab in the Applications Plug in System Settings and add `ibus-daemon -drx` to the startup as a custom command. After this, reboot your system to take effect. Now IBus daemon is automatically started on the system startup.

**Comment by mkniloy** [comment]
done it. still can't write in bijoyClassic or bijoyUnicode...

**Comment by ryonakano** [comment]
Hmm… I wonder what's the difference between your system configuration and mine. 🤔
Could you tell me what keyboard layout did you choose during the installation so that I (possibly) be able to reproduce your system configuration?

**Comment by mkniloy** [comment]
I chose English(US)

**Comment by ryonakano** [comment]
OK, I set up elementary OS Juno and chose English (US) keyboard layout during installation. After that, I set up bijoyClassic and bijoyUnicode. Indeed, I could not type Bengali. However, after executing the following command perhaps I was able to type Bengali (I don't understand Bengali, so I don't know whether characters I typed were really Bengali):

![VirtualBox_eOS in Bengali_07_08_2019_13_20_58](https://user-images.githubusercontent.com/26003928/62594914-48e71180-b917-11e9-90e4-de38111f0156.png)

Could you copy the following command, open Terminal, paste the copied text, and run the command?

```bash
gsettings set org.gnome.desktop.input-sources sources "[('xkb', 'us'), ('ibus', 'm17n:bn:bijoyUnicode'), ('ibus', 'm17n:bn:bijoyClassic')]"
```

After that, reboot your system and check whether you can type Bengali with either bijoyUnicode or bijoyClassic.

**Comment by mkniloy** [comment]
yes..it works! But was it a bug that elementary OS needs to fix..?

**Comment by ryonakano** [comment]
Wow, finally it works! Great to hear that. :tada:

> But was it a bug that elementary OS needs to fix..?

Yes, I think so too. I'm not sure this is correct, but I suppose this might be because elementary OS does not sync the list of input method engines in IBus Preferences and GSettings one (the command above). At this point, #173 might be related to this issue.

**Comment by ryonakano** [comment]
BTW, the instruction on YouTube seems to have some dangerous operations like changing the permission of directories to 777 (which allows everyone who use your system to read, write, or execute any files in the directories) to copy or edit files. I don't think it's a good idea. Instead, we may want to use `sudo` command when we need to access files in the system directory.

So, I put organized instruction here, hope this will be useful for someone who wants to type Bengali in elementary OS including you:

```bash
# Before running the following commands, make sure to download and extract the "Bijoy Linux" folder.

# Open your Terminal, copy and paste the following commands:
cd Bijoy\ Linux
sudo cp bn-bijoy*.png /usr/share/m17n/icons
sudo cp bn-bijoy*.mim /usr/share/m17n
sudo sh -c 'echo -e "/usr/share/m17n/icons/bn-bijoyClassic.png\n/usr/share/m17n/bn-bijoyClassic.mim\n/usr/share/m17n/icons/bn-bijoyUnicode.png\n/usr/share/m17n/bn-bijoyUnicode.mim" >> /var/lib/dpkg/info/m17n-db.list'
gsettings set org.gnome.desktop.input-sources sources "[('xkb', 'us'), ('ibus', 'm17n:bn:bijoyUnicode'), ('ibus', 'm17n:bn:bijoyClassic')]"

# Add the custom command "ibus-daemon -drx" to the startup command from System Settings→Applications→Startup
# Reboot your system. After rebooting, click "Input Method Settings…" in System Settings→Keyboard to open IBus Preferences and add bijoyClassic and bijoyUnicode in the "Input Method" tab. Now switching bijoyUnicode with Ctrl+Space should allow you to type Bengali.
```

**Comment by mkniloy** [comment]
![Screenshot from 2019-08-14 21 28 50](https://user-images.githubusercontent.com/53095171/63034185-a280aa80-beda-11e9-88ee-e0162b014bcd.png)

I tried to use your method...but it's not working...

**Comment by ryonakano** [comment]
It looks like the name of the downloaded zip file differs from mine. The line `cd Bijoy\ Linux` means just **c**hanging **d**irectory to the extracted folder, so changing the folder name to pass the `cd` command will work. Judging from this screenshot the name of the downloaded zip file looks like "Bijoy Ekushe", so running `cd Bijoy\ Ekushe` should work.

I wonder why the name of the folder differs. Did you download the zip file in the website whose link is in [the YouTube demo video](https://www.youtube.com/watch?v=ihkNsNwTcOw) which you've shared? :thinking:

**Comment by mkniloy** [comment]
No... I downloaded from here - https://www.compromath.com/2017/06/install-bijoy-ekushe-to-linuxubuntu.html?m=1
If you have a website you should create an article or make a YouTube video about how to install bijoy in elementary OS. You'll be internet famous! (But the problem is the method will be different when elementary OS fixes this bug 🤔)

**Comment by ryonakano** [comment]
I downloaded it and compared to the one what I have. It looks like they have the same files as to the icons and .mim files. So as I described in my comment above, just changing the folder name should work. :+1:

BTW I don't have any websites or YouTube channels, but the way to make Bijoy available in elementary OS is exactly what I've [commented here](https://github.com/elementary/switchboard-plug-keyboard/issues/231#issuecomment-519015254).
